flujo/application/core/default_cache_components.py

```
@@ -356,21 +357,32 @@ class ThreadSafeMeter:
+    _lock: _thread.LockType = field(init=False, default_factory=_thread.allocate_lock)
 
     async def add(self, cost_usd: float, prompt_tokens: int, completion_tokens: int) -> None:
-        self.total_cost_usd += cost_usd
-        self.prompt_tokens += prompt_tokens
-        self.completion_tokens += completion_tokens
+        with self._lock:
+            self.total_cost_usd += float(cost_usd)
+            self.prompt_tokens += int(prompt_tokens)
+            self.completion_tokens += int(completion_tokens)
 
     async def guard(self, limits: UsageLimits, step_history: list[object] | None = None) -> None:
```

Ruff (0.14.8)

Unused method argument: `step_history`

(ARG002)